The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8590357951/job/23537793458) [(plain)](https://github.com/rust-lang-ci/rust/commit/7427e2c9999b98a67b0d4c176a9319c92e894e3b/checks/23537793458/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [codegen] tests/codegen/thin-lto.rs ... ok

failures:

---- [codegen] tests/codegen/enum/enum-early-otherwise-branch.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-17/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum/enum-early-otherwise-branch/enum-early-otherwise-branch.ll" "/checkout/tests/codegen/enum/enum-early-otherwise-branch.rs" "--check-prefix=CHECK" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/checkout/tests/codegen/enum/enum-early-otherwise-branch.rs:16:12: error: CHECK: expected string not found in input
 // CHECK: [[SELECT:%.*]] = select
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum/enum-early-otherwise-branch/enum-early-otherwise-branch.ll:7:32: note: scanning from here
define noundef zeroext i1 @foo(ptr noalias nocapture noundef readonly align 4 dereferenceable(8) %lhs, ptr noalias nocapture noundef readonly align 4 dereferenceable(8) %rhs) unnamed_addr #0 {
                               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum/enum-early-otherwise-branch/enum-early-otherwise-branch.ll:7:71: note: possible intended match here
define noundef zeroext i1 @foo(ptr noalias nocapture noundef readonly align 4 dereferenceable(8) %lhs, ptr noalias nocapture noundef readonly align 4 dereferenceable(8) %rhs) unnamed_addr #0 {


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum/enum-early-otherwise-branch/enum-early-otherwise-branch.ll
Check file: /checkout/tests/codegen/enum/enum-early-otherwise-branch.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'enum_early_otherwise_branch.85059ca832c3f33d-cgu.0' 
            2: source_filename = "enum_early_otherwise_branch.85059ca832c3f33d-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128" 
            4: target triple = "x86_64-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable 
            7: define noundef zeroext i1 @foo(ptr noalias nocapture noundef readonly align 4 dereferenceable(8) %lhs, ptr noalias nocapture noundef readonly align 4 dereferenceable(8) %rhs) unnamed_addr #0 { 
check:16'0                                    X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:16'1                                                                           ?                                                                                                                           possible intended match
check:16'0     ~~~~~~~
check:16'0     ~~~~~~~
            9:  %0 = load i32, ptr %lhs, align 4, !range !3, !noundef !4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  %1 = load i32, ptr %rhs, align 4, !range !3, !noundef !4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           11:  %_29.not = icmp eq i32 %0, %1 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           12:  br i1 %_29.not, label %bb7, label %bb5 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:16'0     ~
check:16'0     ~
           14: bb7: ; preds = %start 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~
           15:  %_3 = zext i32 %0 to i64 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
           16:  switch i64 %_3, label %bb6 [ 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           17:  i64 0, label %bb2 
check:16'0     ~~~~~~~~~~~~~~~~~~~
           18:  i64 1, label %bb3 
check:16'0     ~~~~~~~~~~~~~~~~~~~
           19:  i64 2, label %bb4 
check:16'0     ~~~~~~~~~~~~~~~~~~~
           20:  ] 
check:16'0     ~~~
check:16'0     ~
check:16'0     ~
           22: bb6: ; preds = %bb7 
check:16'0     ~~~~~~~~~~~~~~~~~~~~
           23:  unreachable 
check:16'0     ~~~~~~~~~~~~~
check:16'0     ~
check:16'0     ~
           25: bb2: ; preds = %bb7 
check:16'0     ~~~~~~~~~~~~~~~~~~~~
           26:  %2 = getelementptr inbounds i8, ptr %lhs, i64 4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           27:  %3 = getelementptr inbounds i8, ptr %rhs, i64 4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           28:  %_18 = load i32, ptr %2, align 4, !noundef !4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           29:  %_19 = load i32, ptr %3, align 4, !noundef !4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           30:  %4 = icmp eq i32 %_18, %_19 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           31:  br label %bb5 
check:16'0     ~~~~~~~~~~~~~~~
check:16'0     ~
check:16'0     ~
           33: bb3: ; preds = %bb7 
check:16'0     ~~~~~~~~~~~~~~~~~~~~
           34:  %5 = getelementptr inbounds i8, ptr %lhs, i64 4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           35:  %6 = getelementptr inbounds i8, ptr %rhs, i64 4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           36:  %_22 = load i32, ptr %5, align 4, !noundef !4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           37:  %_23 = load i32, ptr %6, align 4, !noundef !4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           38:  %7 = icmp eq i32 %_22, %_23 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           39:  br label %bb5 
check:16'0     ~~~~~~~~~~~~~~~
check:16'0     ~
check:16'0     ~
           41: bb4: ; preds = %bb7 
check:16'0     ~~~~~~~~~~~~~~~~~~~~
           42:  %8 = getelementptr inbounds i8, ptr %lhs, i64 4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           43:  %9 = getelementptr inbounds i8, ptr %rhs, i64 4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           44:  %_26 = load i32, ptr %8, align 4, !noundef !4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45:  %_27 = load i32, ptr %9, align 4, !noundef !4 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46:  %10 = icmp eq i32 %_26, %_27 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           47:  br label %bb5 
check:16'0     ~~~~~~~~~~~~~~~
check:16'0     ~
check:16'0     ~
           49: bb5: ; preds = %start, %bb4, %bb3, %bb2 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  %_0.0.shrunk = phi i1 [ %10, %bb4 ], [ %7, %bb3 ], [ %4, %bb2 ], [ false, %start ] 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  ret i1 %_0.0.shrunk 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~
           52: } 
check:16'0     ~~
check:16'0     ~
check:16'0     ~
           54: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:16'0     ~
check:16'0     ~
           56: !llvm.module.flags = !{!0, !1} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57: !llvm.ident = !{!2} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~
check:16'0     ~
check:16'0     ~
           59: !0 = !{i32 8, !"PIC Level", i32 2} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           61: !2 = !{!"rustc version 1.79.0-nightly (7427e2c99 2024-04-07)"} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           62: !3 = !{i32 0, i32 3} 
check:16'0     ~~~~~~~~~~~~~~~~~~~~~
           63: !4 = !{} 
check:16'0     ~~~~~~~~~
------------------------------------------



```

</details>

--- a/compiler/rustc_mir_transform/src/early_otherwise_branch.rs
+++ b/compiler/rustc_mir_transform/src/early_otherwise_branch.rs
@@ -168,7 +168,8 @@ impl<'tcx> MirPass<'tcx> for EarlyOtherwiseBranch {
                 };
                 (value, targets.target_for_value(value))
             });
-            let eq_targets = SwitchTargets::new(eq_new_targets, opt_data.destination);
+            // The otherwise either is the same target branch or an unreachable.
+            let eq_targets = SwitchTargets::new(eq_new_targets, parent_targets.otherwise());
 
             // Create `bbEq` in example above
             let eq_switch = BasicBlockData::new(Some(Terminator {

--- a/tests/codegen/enum/enum-early-otherwise-branch.rs
+++ b/tests/codegen/enum/enum-early-otherwise-branch.rs
@@ -1,0 +1,26 @@
+//@ compile-flags: -O
+//@ min-llvm-version: 18
+
+#![crate_type = "lib"]
+
+pub enum Enum {
+    A(u32),
+    B(u32),
+    C(u32),
+}
+
+#[no_mangle]
+pub fn foo(lhs: &Enum, rhs: &Enum) -> bool {
+    // CHECK-LABEL: define{{.*}}i1 @foo(
+    // CHECK-NOT: switch
+    // CHECK-NOT: br
+    // CHECK: [[SELECT:%.*]] = select
+    // CHECK-NEXT: ret i1 [[SELECT]]
+    // CHECK-NEXT: }
+    match (lhs, rhs) {
+        (Enum::A(lhs), Enum::A(rhs)) => lhs == rhs,
+        (Enum::B(lhs), Enum::B(rhs)) => lhs == rhs,
+        (Enum::C(lhs), Enum::C(rhs)) => lhs == rhs,
+        _ => false,
+    }
+}

--- a/tests/mir-opt/early_otherwise_branch.opt3.EarlyOtherwiseBranch.diff
+++ b/tests/mir-opt/early_otherwise_branch.opt3.EarlyOtherwiseBranch.diff
@@ -37,7 +37,7 @@
 +         StorageLive(_13);
 +         _13 = Ne(_9, move _12);
 +         StorageDead(_12);
-+         switchInt(move _13) -> [0: bb6, otherwise: bb1];
++         switchInt(move _13) -> [0: bb7, otherwise: bb1];
       }
   
       bb1: {
@@ -95,10 +95,13 @@
       }
   
 -     bb9: {
--         unreachable;
 +     bb6: {
+          unreachable;
++     }
++ 
++     bb7: {
 +         StorageDead(_13);
-+         switchInt(_9) -> [0: bb2, 1: bb3, 2: bb4, otherwise: bb1];
++         switchInt(_9) -> [0: bb2, 1: bb3, 2: bb4, otherwise: bb6];
       }
   }
   

--- a/tests/mir-opt/early_otherwise_branch.opt4.EarlyOtherwiseBranch.diff
+++ b/tests/mir-opt/early_otherwise_branch.opt4.EarlyOtherwiseBranch.diff
@@ -37,7 +37,7 @@
 +         StorageLive(_13);
 +         _13 = Ne(_9, move _12);
 +         StorageDead(_12);
-+         switchInt(move _13) -> [0: bb6, otherwise: bb1];
++         switchInt(move _13) -> [0: bb7, otherwise: bb1];
       }
   
       bb1: {
@@ -95,10 +95,13 @@
       }
   
 -     bb9: {
--         unreachable;
 +     bb6: {
+          unreachable;
++     }
++ 
++     bb7: {
 +         StorageDead(_13);
-+         switchInt(_9) -> [0: bb2, 1: bb3, 2: bb4, otherwise: bb1];
++         switchInt(_9) -> [0: bb2, 1: bb3, 2: bb4, otherwise: bb6];
       }
   }
   

--- a/tests/mir-opt/early_otherwise_branch_3_element_tuple.opt2.EarlyOtherwiseBranch.diff
+++ b/tests/mir-opt/early_otherwise_branch_3_element_tuple.opt2.EarlyOtherwiseBranch.diff
@@ -47,7 +47,7 @@
 +         StorageLive(_19);
 +         _19 = Ne(_14, move _18);
 +         StorageDead(_18);
-+         switchInt(move _19) -> [0: bb9, otherwise: bb1];
++         switchInt(move _19) -> [0: bb10, otherwise: bb1];
       }
   
       bb1: {
@@ -129,10 +129,13 @@
       }
   
 -     bb12: {
--         unreachable;
 +     bb9: {
+          unreachable;
++     }
++ 
++     bb10: {
 +         StorageDead(_19);
-+         switchInt(_14) -> [0: bb2, 1: bb3, 2: bb4, otherwise: bb1];
++         switchInt(_14) -> [0: bb2, 1: bb3, 2: bb4, otherwise: bb9];
       }
   }
   

